lib/middleware/csrfProtected.ts

```diff
@@ -0,0 +1,28 @@
+import { NextApiRequest, NextApiResponse } from "next";
+import { MiddlewareRequest, MiddlewareReturn } from "@lib/types";
+import { getCsrfToken } from "next-auth/client";
+import { logMessage } from "@lib/logger";
+
+export const csrfProtected = (allowMethods: string[]): MiddlewareRequest => {
+  return async (req: NextApiRequest, res: NextApiResponse): Promise<MiddlewareReturn> => {
+    try {
+      if (allowMethods && req.method && allowMethods.includes(req.method)) {
+        const csrfToken = await getCsrfToken({ req });
+        if (!csrfToken) throw Error("Invalid Csrf Token found");
+        if (csrfToken === req.headers["x-csrf-token"]) {
+          return { next: true };
+        } else {
+          res.status(403).json({ error: "Access Denied" });
+          return { next: false };
+        }
+      } else {
```

Comment:
If the method is not CSRF protected then it should pass through and not throw a "Method is not Allowed".  The CORS middleware already checks for which methods the API endpoint will accept.  Similar to the 'sessionExists' middleware it should only apply it's logic to specific declared methods.